streams/src/main/java/org/apache/kafka/streams/state/internals/Segments.java

```diff
@@ -121,89 +130,66 @@ void openExisting(final ProcessorContext context) {
                     throw new ProcessorStateException(String.format("dir %s doesn't exist and cannot be created for segments %s", dir, name));
                 }
             }
-        } catch (Exception ex) {
+        } catch (final Exception ex) {
             // ignore
         }
+
+        final long minLiveSegment = segmentId(context.streamTime() - retentionPeriod);
+        cleanupEarlierThan(minLiveSegment);
     }
 
     List<Segment> segments(final long timeFrom, final long timeTo) {
-        final long segFrom = Math.max(minSegmentId, segmentId(Math.max(0L, timeFrom)));
-        final long segTo = Math.min(maxSegmentId, segmentId(Math.min(maxSegmentId * segmentInterval, Math.max(0, timeTo))));
-
-        final List<Segment> segments = new ArrayList<>();
-        for (long segmentId = segFrom; segmentId <= segTo; segmentId++) {
-            Segment segment = getSegment(segmentId);
-            if (segment != null && segment.isOpen()) {
-                try {
-                    segments.add(segment);
-                } catch (InvalidStateStoreException ise) {
-                    // segment may have been closed by streams thread;
-                }
+        final List<Segment> result = new ArrayList<>();
```

Comment:
This is not necessarily part of this PR, but right now we only purge segments in `put`, should we consider doing purge in `segments` and `allSegments()` as well? They cover all range queries, i.e. both read and writes can trigger segment purging.

Reply:
We could do that, but it might be better to only do it in `put`.

The disk usage can only grow during `put`, so there's no danger of running out of disk if you garbage collect only during `put`. Collecting in extra places would just be an optimization to keep the disk footprint as small as possible.

Also, it might be surprising to perform disk i/o during a metadata call to list segments.

That said, it wouldn't be incorrect to do it...

Reply:
That is a good point. Let's keep our current behavior as only purge on puts then so that range queries do not see surprise perf variance due to rocksdb closing (each may take a few seconds in worst case).